airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py

```diff
@@ -0,0 +1,142 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+import concurrent.futures
+import datetime
+import math
+from abc import ABC
+from dataclasses import asdict
+from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+from .api import ZohoAPI
+from .exceptions import IncompleteMetaDataException, UnknownDataTypeException
+from .types import FieldMeta, ModuleMeta, ZohoPickListItem
+
+
+class ZohoCrmStream(HttpStream, ABC):
+    primary_key: str = "id"
+    module: ModuleMeta = None
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if response.status_code != 200:
+            return None
+        pagination = response.json()["info"]
+        if not pagination["more_records"]:
+            return None
+        return {"page": pagination["page"] + 1}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return next_page_token or {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        data = response.json()["data"] if response.status_code == 200 else []
+        yield from data
+
+    def path(self, *args, **kwargs) -> str:
+        return f"/crm/v2/{self.module.api_name}"
+
+    def get_json_schema(self) -> Optional[Dict[Any, Any]]:
+        try:
+            return asdict(self.module.schema)
+        except IncompleteMetaDataException:
+            self.logger.warning(f"Could not retrieve fields Metadata for module {self.module.api_name}, skipping")
```

Comment:
Shouldn’t this throw an exception?

Reply:
Same for branch below

Reply:
As for branch below - changed it to reraise exception.
As for this one - it happens when we are unable to construct stream schema due to API response. If we reraise exception - `discover` and `read` will fail. If we return None - current stream will be skipped in both `discover` and `read`, but others will work.

I added a comment on how schema generation works so the reader can understand why this exception may be raised and should not be reraised further.
```
            # to build a schema for a stream, a sequence of requests is made:
            # one `/settings/modules` which introduces a list of modules,
            # one `/settings/modules/{module_name}` per module and
            # one `/settings/fields?module={module_name}` per module.
            # Any of former two can result in 204 and empty body what blocks us
            # from generating stream schema and, therefore, a stream.
```

Reply:
I see, so to be clear, receiving a 204 constitutes a valid case where we can skip syncing this stream?

Reply:
yes, correct